# Práctica Machine Learning KC

## Problema de regresión para predecir el precio de un airbnb 

En este notebook se revisarán los conceptos de:



## 1. Carga de datos, limpieza y división train/test 

En primer lugar, se va a echar un vistazo previo a los datos, eliminando todos aquellos campos que no nos interesen como URLs, reseñas...

Así pues, se han eliminado lo ssiguientes campos del data set original, que se consideran que no aportan valor:
* Listing Url
* Scrape ID
* Last Scraped
* Name (con el ID es suficiente)
* Summary
* Space
* Description
* Experience
* Neighborhood Overview
* Notes
* Transit
* Access
* Interaction 
* House Rules
* Thumbnail Url	
* Medium Url	
* Picture Url	
* XL Picture Url
* Host ID
* Host URL
* Host Name
* Host Since
* Host Location
* Host About
* Host Response Time
* Host Response Rate
* Host Acceptance Rate
* Host Thumbnail Url	
* Host Picture Url	
* Host Neighbourhood
* Host Listings Count	
* Host Total Listings Count	
* Host Verifications
* Street
* Neighbourhood	
* Neighbourhood Cleansed
* State	
* Zipcode	
* Market	
* Smart Location	
* Country Code	
* Country	
* Latitude	
* Longitude
* Bed Type
* Square Feet
* Weekly Price	
* Monthly Price	
* Security Deposit	
* Cleaning Fee	
* Guests Included
* Extra People	
* Minimum Nights
* Maximum Nights
* Calendar Updated	
* Has Availability	
* Availability 30	
* Availability 60	
* Availability 90	
* Availability 365	
* Calendar last Scraped
* First Review	
* Last Review
* Review Scores Accuracy	
* Review Scores Cleanliness	
* Review Scores Checkin	
* Review Scores Communication	
* Review Scores Location	
* Review Scores Value	
* License	Jurisdiction Names
* Cancellation Policy	
* Calculated host listings count

Así pues, nos quedaremos con los siguientes atributos:


| Atributo              | Descripción                                      |
| :-                    | :-                                               |
| *City*                | Ciudad donde se encuentra la vivienda            |
| *Property Type*       | Tipo de propiedad (loft, apartamento, etc.)      |
| *Room Type*           | Tipo de habitación (entera, privada, compartida) |
| *Accommodates*        | Número de personas que puede alojar              |
| *Bathrooms*           | Número de baños/aseos                            |
| *Bedrooms*            | Número de habitaciones                           |
| *Beds*                | Número de camas                                  |
| *Bed Type*            | Tipo de cama (individual, doble, etc.)           |
| *Amenities*           | Servicios y comodidades disponibles              |
| *Number of Reviews*   | Número de reseñas                                |
| *Review Scores Rating*| Puntuación de las reseñas                        |
| *Reviews per Month*   | Número de reseñas por mes                        |
| *Geolocation*         | Ubicación geográfica (latitud y longitud)        |
| *Features*            | Características adicionales                      |
| *Price*               | Precio de la vivienda                            |

A continuación, procedemos a cargar los datos para proceder con su visualización:


In [10]:
# Importamos las librerias necesarias
import numpy  as np
import pandas as pd
from sklearn.model_selection import train_test_split
import matplotlib.pyplot as plt

%matplotlib inline

In [36]:
house_data = pd.read_csv("./data/airbnb-listings-extract-cleaned.csv", sep=';', encoding = "latin")
print(house_data.shape)
house_data.head(5).T  

(14780, 16)


0  \
ID                                                             11210388   
City                                                             Austin   
Property Type                                                      Loft   
Room Type                                               Entire home/apt   
Accommodates                                                          2   
Bathrooms                                                          10.0   
Bedrooms                                                            1.0   
Beds                                                                1.0   
Bed Type                                                       Real Bed   
Amenities             TV,Cable TV,Internet,Wireless Internet,Air con...   
Number of Reviews                                                    42   
Review Scores Rating                                               98.0   
Reviews per Month                                                  35.0   
Geolocation                                30.3373609355,-97.8632766782   
Features              Host Is Superhost,Host Has Profile Pic,Host Id...   
Price                                                              58.0   

                                                                      1  \
ID                                                             17471131   
City                                                          Barcelona   
Property Type                                                 Apartment   
Room Type                                               Entire home/apt   
Accommodates                                                          4   
Bathrooms                                                          10.0   
Bedrooms                                                            1.0   
Beds                                                                1.0   
Bed Type                                                       Real Bed   
Amenities             Wireless Internet,Air conditioning,Elevator in...   
Number of Reviews                                                     1   
Review Scores Rating                                               80.0   
Reviews per Month                                                  86.0   
Geolocation                                 41.3896829422,2.17262543017   
Features              Host Has Profile Pic,Requires License,Instant ...   
Price                                                              70.0   

                                                                      2  \
ID                                                             17584891   
City                                                          Barcelona   
Property Type                                                 Apartment   
Room Type                                               Entire home/apt   
Accommodates                                                          4   
Bathrooms                                                          10.0   
Bedrooms                                                            1.0   
Beds                                                                1.0   
Bed Type                                                       Real Bed   
Amenities             Wireless Internet,Air conditioning,Elevator in...   
Number of Reviews                                                     0   
Review Scores Rating                                                NaN   
Reviews per Month                                                   NaN   
Geolocation                                 41.3930345489,2.16217327868   
Features              Host Has Profile Pic,Requires License,Instant ...   
Price                                                              70.0   

                                                                      3  \
ID                                                              5398030   
City                                                          Barcelona   
Property Type    

Visualizando los datos se observan varios problemas que pueden afectar al entrenamiento de los modelos. 

En primer lugar se detecta un número inusualmente de baños, ya que no tiene mucho sentido que un apartamento con capacidad para 2 personas tenga 10 baños. Esto puede deberse a un error en la carga de datos, por eso, se ha confirmado que al importar el csv con excel para eliminar las columnas no deseadas ha habido un error. En el dataset original, el atributo bathrooms es de tipo float e incluía un punto que separaba la parte entera de la decimal. Al importar los datos se ha eliminado el punto que los separaba, por lo que se va a dividir entre 10 todos los valores de este atributo.

In [37]:
house_data['Bathrooms'] = house_data['Bathrooms'] / 10
house_data.head(5).T  

0  \
ID                                                             11210388   
City                                                             Austin   
Property Type                                                      Loft   
Room Type                                               Entire home/apt   
Accommodates                                                          2   
Bathrooms                                                           1.0   
Bedrooms                                                            1.0   
Beds                                                                1.0   
Bed Type                                                       Real Bed   
Amenities             TV,Cable TV,Internet,Wireless Internet,Air con...   
Number of Reviews                                                    42   
Review Scores Rating                                               98.0   
Reviews per Month                                                  35.0   
Geolocation                                30.3373609355,-97.8632766782   
Features              Host Is Superhost,Host Has Profile Pic,Host Id...   
Price                                                              58.0   

                                                                      1  \
ID                                                             17471131   
City                                                          Barcelona   
Property Type                                                 Apartment   
Room Type                                               Entire home/apt   
Accommodates                                                          4   
Bathrooms                                                           1.0   
Bedrooms                                                            1.0   
Beds                                                                1.0   
Bed Type                                                       Real Bed   
Amenities             Wireless Internet,Air conditioning,Elevator in...   
Number of Reviews                                                     1   
Review Scores Rating                                               80.0   
Reviews per Month                                                  86.0   
Geolocation                                 41.3896829422,2.17262543017   
Features              Host Has Profile Pic,Requires License,Instant ...   
Price                                                              70.0   

                                                                      2  \
ID                                                             17584891   
City                                                          Barcelona   
Property Type                                                 Apartment   
Room Type                                               Entire home/apt   
Accommodates                                                          4   
Bathrooms                                                           1.0   
Bedrooms                                                            1.0   
Beds                                                                1.0   
Bed Type                                                       Real Bed   
Amenities             Wireless Internet,Air conditioning,Elevator in...   
Number of Reviews                                                     0   
Review Scores Rating                                                NaN   
Reviews per Month                                                   NaN   
Geolocation                                 41.3930345489,2.16217327868   
Features              Host Has Profile Pic,Requires License,Instant ...   
Price                                                              70.0   

                                                                      3  \
ID                                                              5398030   
City                                                          Barcelona   
Property Type    

Por lo demás no se observan más errores en la carga, por lo que pasamos a dividir los conjuntos de train y test antes de la imputación y limpieza de datos. Esto se realiza antes para evitar un posible sesgo en los datos, para tener una correcta validación al tener un conjunto de datos de test intactos antes de la evaluación del rendimiento del modelo y una simulación de valores reales. 

Para la separación de los conjuntos se va a utilizar la fuinción `train_test_split` de sklearn. Adicionalmente se va a dividir un 80% en el conjunto de train y un 20% para el conjunto de test, y se va a realizar el suffle para mezclar los datos y evitar el sesgo en caso de que los datos pudieran estar ordenados.


In [39]:
train, test = train_test_split(house_data, test_size=0.2, shuffle=True, random_state=0)

print(f'Dimensiones del dataset de training: {train.shape}')
print(f'Dimensiones del dataset de test: {test.shape}')

# Guardamos
train.to_csv('./data/house_data_train.csv', sep=';', decimal='.', index=False)
test.to_csv('./data/house_data_test.csv', sep=';', decimal='.', index=False)

# A partir de este momento cargamos el dataset de train y trabajamos ÚNICAMENTE con él. 

house_data = pd.read_csv('./data/house_data_train.csv', sep=';', decimal='.')
house_data.head(5).T

Dimensiones del dataset de training: (11824, 16)
Dimensiones del dataset de test: (2956, 16)


0  \
ID                                                              5994463   
City                                                             Madrid   
Property Type                                                 Apartment   
Room Type                                               Entire home/apt   
Accommodates                                                          4   
Bathrooms                                                           1.0   
Bedrooms                                                            1.0   
Beds                                                                2.0   
Bed Type                                                       Real Bed   
Amenities             TV,Internet,Wireless Internet,Air conditioning...   
Number of Reviews                                                    12   
Review Scores Rating                                               95.0   
Reviews per Month                                                   5.0   
Geolocation                                40.4077318793,-3.68481869733   
Features              Host Has Profile Pic,Is Location Exact,Require...   
Price                                                              60.0   

                                                                      1  \
ID                                                             14136180   
City                                                             Madrid   
Property Type                                                 Apartment   
Room Type                                               Entire home/apt   
Accommodates                                                          4   
Bathrooms                                                           1.0   
Bedrooms                                                            1.0   
Beds                                                                2.0   
Bed Type                                                       Real Bed   
Amenities             TV,Internet,Wireless Internet,Air conditioning...   
Number of Reviews                                                    20   
Review Scores Rating                                               91.0   
Reviews per Month                                                 243.0   
Geolocation                                40.4158022422,-3.70534037765   
Features              Host Has Profile Pic,Host Identity Verified,Re...   
Price                                                              50.0   

                                                                      2  \
ID                                                             15520134   
City                                                             Madrid   
Property Type                                                 Apartment   
Room Type                                                   Shared room   
Accommodates                                                          1   
Bathrooms                                                           1.5   
Bedrooms                                                            1.0   
Beds                                                                8.0   
Bed Type                                                       Real Bed   
Amenities             Internet,Wireless Internet,Wheelchair accessib...   
Number of Reviews                                                     0   
Review Scores Rating                                                NaN   
Reviews per Month                                                   NaN   
Geolocation                                40.3890481626,-3.74037392557   
Features              Host Has Profile Pic,Host Identity Verified,Is...   
Price                                                              10.0   

                                                                      3  \
ID                                                              8809721   
City                                                             Madrid   
Property Type    

## 2. Análisis exploratorio

Vamos a proceder al análisis de los datos viendo usando funciones de Pandas para intentar sacar conclusiones a cerca de los datos.

In [40]:
house_data.describe()

ID  Accommodates     Bathrooms     Bedrooms          Beds  \
count  1.182400e+04  11824.000000  11780.000000  11804.00000  11787.000000   
mean   1.025308e+07      3.277486      1.285229      1.34429      2.049122   
std    5.566485e+06      2.093973      0.664691      0.90518      1.623489   
min    1.986400e+04      1.000000      0.000000      0.00000      1.000000   
25%    5.510692e+06      2.000000      1.000000      1.00000      1.000000   
50%    1.125754e+07      3.000000      1.000000      1.00000      2.000000   
75%    1.531982e+07      4.000000      1.000000      2.00000      2.000000   
max    1.858361e+07     16.000000      8.000000     10.00000     16.000000   

       Number of Reviews  Review Scores Rating  Reviews per Month  \
count       11824.000000           9163.000000        9285.000000   
mean           22.664834             91.628179         163.085407   
std            38.092338              9.137614         183.128112   
min             0.000000             20.000000           1.000000   
25%             1.000000             89.000000          29.000000   
50%             7.000000             94.000000          88.000000   
75%            27.000000             98.000000         243.000000   
max           356.000000            100.000000        1721.000000   

              Price  
count  11809.000000  
mean      73.712592  
std       71.624844  
min        9.000000  
25%       34.000000  
50%       55.000000  
75%       87.000000  
max      969.000000

El método `describe()` nos arroja los principales valores estadísticos y nos permite sacar las siguientes conclusiones:
1. Vemos que en el conteo de atributos no todas las columnas tienen el valor 11824, por lo que va a ver atributos nulos. Más adelante se decidirá qué hacer con ellos.
2. La fila MAX nos arroja información que puede tratarse de outliers. Por ejemplo, vemos que algún apartamento recibe 1721 reviews al mes, lo cual a priori parece algo elevado, aunque puede ser real ya que suponiendo que tiene una capacidad para 16 huéspedes (el máximo) si cada día cambiara de huéspedes y todo el mes alquilara las 16 plazas, suponiendo que todos los huéspedes dejan una reseña, podría tener como máximo 16 * 30 =  480 reseñas en un mes. 

Para saber qué atributos tienen valores nulos vamos a ejecutar las funciones `isnull().any()`:

In [38]:
house_data.isnull().any()

ID                      False
City                     True
Property Type           False
Room Type               False
Accommodates            False
Bathrooms                True
Bedrooms                 True
Beds                     True
Bed Type                False
Amenities                True
Number of Reviews       False
Review Scores Rating     True
Reviews per Month        True
Geolocation             False
Features                 True
Price                    True
dtype: bool

En función del atributo, vamos a completar los valores nulos con la media o con 0 para valores numéricos y con Unknown en caso de atributos de tipo texto.

### Media
* Bathrooms
* Bedrooms
* Beds
* Price. En este caso nos interesa que sea la media ya que es nuestra variable objetivo.

### Valor 0
* Review Scores Rating
* Reviews per Month 

### Unknown 
* City 
* Amenities
* Features

In [43]:
house_data['Bathrooms'].fillna(house_data['Bathrooms'].mean(), inplace=True)
house_data['Bedrooms'].fillna(house_data['Bedrooms'].mean(), inplace=True)
house_data['Beds'].fillna(house_data['Beds'].mean(), inplace=True)
house_data['Price'].fillna(house_data['Price'].mean(), inplace=True)

house_data['Review Scores Rating'].fillna(house_data['Review Scores Rating'].mode()[0], inplace=True)
house_data['Reviews per Month'].fillna(house_data['Reviews per Month'].mode()[0], inplace=True)

house_data['City'].fillna('Unknown', inplace=True)
house_data['Amenities'].fillna('Unknown', inplace=True)
house_data['Features'].fillna('Unknown', inplace=True)

house_data.isnull().any()
house_data.head(5).T

0  \
ID                                                              5994463   
City                                                             Madrid   
Property Type                                                 Apartment   
Room Type                                               Entire home/apt   
Accommodates                                                          4   
Bathrooms                                                           1.0   
Bedrooms                                                            1.0   
Beds                                                                2.0   
Bed Type                                                       Real Bed   
Amenities             TV,Internet,Wireless Internet,Air conditioning...   
Number of Reviews                                                    12   
Review Scores Rating                                               95.0   
Reviews per Month                                                   5.0   
Geolocation                                40.4077318793,-3.68481869733   
Features              Host Has Profile Pic,Is Location Exact,Require...   
Price                                                              60.0   

                                                                      1  \
ID                                                             14136180   
City                                                             Madrid   
Property Type                                                 Apartment   
Room Type                                               Entire home/apt   
Accommodates                                                          4   
Bathrooms                                                           1.0   
Bedrooms                                                            1.0   
Beds                                                                2.0   
Bed Type                                                       Real Bed   
Amenities             TV,Internet,Wireless Internet,Air conditioning...   
Number of Reviews                                                    20   
Review Scores Rating                                               91.0   
Reviews per Month                                                 243.0   
Geolocation                                40.4158022422,-3.70534037765   
Features              Host Has Profile Pic,Host Identity Verified,Re...   
Price                                                              50.0   

                                                                      2  \
ID                                                             15520134   
City                                                             Madrid   
Property Type                                                 Apartment   
Room Type                                                   Shared room   
Accommodates                                                          1   
Bathrooms                                                           1.5   
Bedrooms                                                            1.0   
Beds                                                                8.0   
Bed Type                                                       Real Bed   
Amenities             Internet,Wireless Internet,Wheelchair accessib...   
Number of Reviews                                                     0   
Review Scores Rating                                              100.0   
Reviews per Month                                                  10.0   
Geolocation                                40.3890481626,-3.74037392557   
Features              Host Has Profile Pic,Host Identity Verified,Is...   
Price                                                              10.0   

                                                                      3  \
ID                                                              8809721   
City                                                             Madrid   
Property Type    